pyanaconda/modules/payloads/payload/flatpak/source.py

```diff
@@ -0,0 +1,435 @@
+                    result.append(StaticSourceImage(digest, manifest_json, config_json))
+
+        return result
+
+    def _blob_url(self, digest):
+        if not digest.startswith("sha256:"):
+            raise RuntimeError("Only SHA-256 digests are supported")
+        return self._url + "/blobs/sha256/" + digest[7:]
+
+    def _get_blob(self, downloader, digest) -> bytes:
+        result = self._cached_blobs.get(digest)
+        if result:
+            return result
+
+        response = downloader(self._blob_url(digest))
+        response.raise_for_status()
+
+        self._cached_blobs[digest] = result = response.content
+        return result
+
+    def _download_blob(self, downloader, download_location, digest, stream=False):
+        if not digest.startswith("sha256:"):
+            raise RuntimeError("Only SHA-256 digests are supported")
+
+        blobs_dir = os.path.join(download_location, "blobs/sha256/")
+        os.makedirs(blobs_dir, exist_ok=True)
+
+        path = os.path.join(blobs_dir, digest[7:])
+        with open(path, "wb") as f:
+            if stream:
+                response = downloader(self._blob_url(digest), stream=True)
+                response.raise_for_status()
+                size = 0
+                while True:
+                    chunk = response.raw.read(64*1024)
+                    if not chunk:
+                        break
+                    size += len(chunk)
+                    f.write(chunk)
+                return size
+            else:
+                blob = self._get_blob(downloader, digest)
+                f.write(blob)
+                return len(blob)
+
+    def _get_json(self, session, digest):
+        return json.loads(self._get_blob(session, digest))
+
+
+class RegistrySourceImage(SourceImage):
+    def __init__(self, labels):
+        self._labels = labels
+
+    @property
+    def labels(self):
+        return self._labels
+
+
+class FlatpakRegistrySource(FlatpakSource):
+    """Flatpak images indexed by a remote JSON file, and stored in a registry.
+
+    https://github.com/flatpak/flatpak-oci-specs/blob/main/registry-index.md
+    """
+
+    def __init__(self, url):
+        self._index = None
+        self._url = url
+
+    def calculate_size(self, refs):
+        # For registry sources, we don't download the images in advance;
+        # instead they are downloaded into the /var/tmp of the target
+        # system and installed one-by-one. So the downloads don't count
+        # towards the space in the temporary download location, but we
+        # need space for the largest download in the target system.
+        # (That space will also be needed for upgrades after installation.)
+
+        log.debug("Calculating size of: %s", refs)
+
+        max_download_size = 0
+        installed_size = 0
+        expanded = self._expand_refs(refs)
+
+        for image in self._images:
+            if image.ref not in expanded:
```

Comment:
Again, please note why we don't care about expanded refs here.

Reply:
Fixed.